test/fail_compilation/test23058.c

```diff
@@ -0,0 +1,23 @@
+/* TEST_OUTPUT:
+---
+fail_compilation/test23058.c(105): Error: array index 4 is out of bounds `[0..3]`
+fail_compilation/test23058.c(110): Error: array index 5 is out of bounds `[0..4]`
+---
+ */
+
+/* https://issues.dlang.org/show_bug.cgi?id=23058
+ */
+
+#line 100
+
+int arr[3][4] = { { 1,2,3,4 }, { 5,6,7,8 }, { 9,10,11,12 } };
+
+void test1()
+{
+    int *p1 = &arr[4][2];
```

Comment:
This code is technically correct code from the perspective of the C specification. Is it ok to turn it into an error?

Reply:
It's length+1, which is not technically correct as I recall.

Reply:
I compiled this code with gcc and there's no error. Also, the actual code [you are introducing](https://github.com/dlang/dmd/pull/14094/files#diff-be3f929dc15535ccf5bbb3d9f274993cac59cad6f83105dda698b28338bbb323R571) treats C and D code the same.

Reply:
You're right that gcc doesn't give an error for:
```
int a[10];
int* f() { return &a[100]; }
```
but clang does with `-Wpedantic -std=c11`
The C11 6.5.6-8 spec lists such constructs as "undefined behavior". I'm good with diagnosing an error at compile time.

Reply:
There's no reason it should be different for D.